libraries/lib-numeric-formats/NumericConverter.cpp

```diff
@@ -70,418 +53,59 @@ static const TranslatableString BuildBeatsFormat() {
-      XO("thousandths of octaves")
-   }
-   },
-
-   {
-   /* i18n-hint: Name of display format that shows log of frequency
-    * in semitones and cents */
-   { XO("semitones + cents") },
-   {
-      /* i18n-hint: Format string for displaying log of frequency in semitones
-       * and cents.
-       * Change the decimal points for your locale. Don't change the numbers.
-       * The decimal separator is specified using '<' if your language uses a ',' or
-       * to '>' if your language uses a '.'. */
-      XO("1000 semitones >0100 cents|17.312340491"),   // Scale factor is 12 / ln (2)
-      /* i18n-hint: a cent is a hundredth of a semitone (which is 1/12 octave) */
-      XO("hundredths of cents")
-   }
-   },
-   
-   {
-   /* i18n-hint: Name of display format that shows log of frequency
-    * in decades */
-   { XO("decades") },
-   {
-      /* i18n-hint: Format string for displaying log of frequency in decades.
-       * Change the decimal points for your locale. Don't change the numbers. */
-      XO("10>01000 decades|0.434294482"),   // Scale factor is 1 / ln (10)
-      /* i18n-hint: a decade is a tenfold increase of frequency */
-      XO("thousandths of decades")
-   }
-   },
-};
-
-   const BuiltinFormatString *ChooseBuiltinFormatStrings
-      (NumericConverterType type)
-   {
-      switch (type) {
-         default:
-      case NumericConverterType::TIME:
-            return TimeConverterFormats_;
-      case NumericConverterType::FREQUENCY:
-            return FrequencyConverterFormats_;
-      case NumericConverterType::BANDWIDTH:
-            return BandwidthConverterFormats_;
-      }
-   }
-
-   size_t ChooseNBuiltinFormatStrings
-      (NumericConverterType type)
-   {
-      switch (type) {
-         default:
-      case NumericConverterType::TIME:
-            return WXSIZEOF(TimeConverterFormats_);
-      case NumericConverterType::FREQUENCY:
-            return WXSIZEOF(FrequencyConverterFormats_);
-      case NumericConverterType::BANDWIDTH:
-            return WXSIZEOF(BandwidthConverterFormats_);
-      }
-   }
-}
-
 //
 // ----------------------------------------------------------------------------
 // NumericConverter Class
 // ----------------------------------------------------------------------------
 //
 NumericFormatSymbol NumericConverter::DefaultSelectionFormat()
-{ return TimeConverterFormats_[5].name; }
+{ return L"hh:mm:ss + milliseconds"; }
 NumericFormatSymbol NumericConverter::TimeAndSampleFormat()
-{ return TimeConverterFormats_[6].name; }
+{ return L"hh:mm:ss + samples"; }
 NumericFormatSymbol NumericConverter::SecondsFormat()
-{ return TimeConverterFormats_[0].name; }
+{ return L"seconds"; }
 NumericFormatSymbol NumericConverter::HoursMinsSecondsFormat()
-{ return TimeConverterFormats_[1].name; }
+{ return L"hh:mm:ss"; }
 NumericFormatSymbol NumericConverter::HundredthsFormat()
-{ return TimeConverterFormats_[3].name; }
+{ return L"hh:mm:ss + hundredths"; }
```

Comment:
The introduction of long-distance coincidence of strings literals between two source files is unfortunate.

I don't see yet how these functions could be moved into ParsedNumericConverterFormatter.cpp

Reply:
Why should they?

Reply:
No -- that's wrong.  With only one exception (DefaultSelectionFormat), they are not used within this file, but are only used by various dialogs and toolbars.  So they easily can and should be moved to that other file.

What to do about DefaultSelectionFormat to make it nicer, I don't know yet.

Reply:
> So they easily can and should be moved to that other file

I do agree with this. I've just tried to keep this review 1% less long and boring

Reply:
As an option I can move this specific formats into the NumericConverter.cpp, so I can keep all the constants in one file. This will, however, affect the user visible ordering. I don't think that this is desirable.

Reply:
> > So they easily can and should be moved to that other file
> 
> I do agree with this. I've just tried to keep this review 1% less long and boring

You do -- good, please do the small cut and paste.  Seeing well organized code pleases me rather than bores me.

For the same reason we like `BoolSetting` and similar for eliminating repetition of a string literal in favor of just one named object -- so this would be good.

Reply:
> As an option I can move this specific formats into the NumericConverter.cpp, so I can keep all the constants in one file. This will, however, affect the user visible ordering. I don't think that this is desirable.

I would not like that.

Really I think time and frequency formats would be nicer separated into their own files, but let's not do that yet.

Reply:
> For the same reason we like BoolSetting

Introducing new (bad) unintended dependencies by doing that, as we saw in this PR already. That specific case should be solved differently though, we should instead have listeners for the preferences reset.  

TBH I'm quite confused by how you don't like enums, but instead replacing them with constants or global objects :-)

Reply:
Speaking of preferences objects - personally I see shared global writable preferences as a massive source of potential bugs. Using objects doesn't make it worse, but it doesn't really fix a problem either.

Reply:
> > For the same reason we like BoolSetting
> 
> Introducing new (bad) unintended dependencies by doing that, as we saw in this PR already. That specific case should be solved differently though, we should instead have listeners for the preferences reset.
> 
> TBH I'm quite confused by how you don't like enums, but instead replacing them with constants or global objects :-)

I see the Settings object as revealing the need to make a link dependency explicit.  The problems of bad global state may not be solved but the naming an object with linkage, instead of the reliance on coincident string literals, forces a more deliberate choice about where to define the global object and which place to make dependent on the other place.

Reply:
I dislike enums for a different reason.  It's closed-endedness.  If instead the set of possibilities may need to grow and serve very unrelated special cases, I would rather have a registry and also good dependency inversion.